core/renderers/block_rendering_rewrite/measurables.js

```diff
@@ -419,78 +436,14 @@ Blockly.blockRendering.RoundCorner = function() {
 };
 goog.inherits(Blockly.blockRendering.RoundCorner, Blockly.blockRendering.Measurable);
 
-
-Blockly.blockRendering.Row = function() {
-  this.type = 'row';
-  this.yPos = 0;
-  this.elements = [];
-  this.width = 0;
-  this.height = 0;
-  this.widthWithConnectedBlocks = 0;
-
-  this.hasExternalInput = false;
-  this.hasStatement = false;
-  this.hasInlineInput = false;
-  this.hasDummyInput = false;
-  this.hasJaggedEdge = false;
-};
-
 /**
- * The shape object to use when drawing previous and next connections.
- * TODO (#2803): Formalize type annotations for these objects.
- * @type {Object}
+ * An object containing information about a spacer between two rows.
+ * @param {number} height The height of the spacer.
+ * @param {number} width The width of the spacer.
+ * @package
+ * @constructor
+ * @extends {Blockly.blockRendering.Measurable}
  */
-Blockly.blockRendering.Row.prototype.notchShape =
-    Blockly.blockRendering.constants.NOTCH;
-
-Blockly.blockRendering.Row.prototype.isSpacer = function() {
-  return false;
-};
-
-Blockly.blockRendering.Row.prototype.measure = function() {
-  var connectedBlockWidths = 0;
-  for (var e = 0; e < this.elements.length; e++) {
-    var elem = this.elements[e];
-    this.width += elem.width;
-    if (elem.isInput) {
-      if (elem.type == 'statement input') {
-        connectedBlockWidths += elem.connectedBlockWidth;
-      } else if (elem.type == 'external value input' &&
-          elem.connectedBlockWidth != 0) {
-        connectedBlockWidths += (elem.connectedBlockWidth - elem.connectionWidth);
-      }
-    }
-    if (!(elem.isSpacer())) {
-      this.height = Math.max(this.height, elem.height);
-    }
-  }
-  this.widthWithConnectedBlocks = this.width + connectedBlockWidths;
-};
-
-Blockly.blockRendering.Row.prototype.getLastInput = function() {
-  for (var i = this.elements.length - 1; i >= 0; i--) {
-    var elem = this.elements[i];
-    if (elem.isSpacer()) {
-      continue;
-    }
-    if (elem.isInput) {
-      return elem;
-    } else if (elem.isField()) {
-      return elem.parentInput;
-    }
-  }
-  // Return null if there are no inputs.
-  return null;
-};
-
-Blockly.blockRendering.Row.prototype.getFirstSpacer = function() {
-  return this.elements[0];
-};
-
-Blockly.blockRendering.Row.prototype.getLastSpacer = function() {
-  return this.elements[this.elements.length - 1];
-};
-
 Blockly.blockRendering.BetweenRowSpacer = function(height, width) {
```

Comment:
I prefer the between row spacer to inherit from Row, because of its usage as an element in the rows array in block_render_info. There's also a few instances in code where it is referred to as a row, like in makeSpacerRow_ (the name implies it is a row and the jsdoc refers to it as a row).

Reply:
Agreed.  I just did some poking and my current proposal is:
- Row is the parent class
- There are four subclasses:
  - TopRow
  - BottomRow
  - SpacerRow
  - InputRow
- No one ever instantiates a Row directly (always a subclass)
- SpacerRow contains an array of elements, which contains one InRowSpacer.  This gets rid of the need to check if a row is a spacer row before iterating across its element array.

This lets me make many of the annotations tighter (albeit with some casts) and removes some unnecessary distinctions between spacer and non spacer rows.